dozer-ingestion/README.md

```diff
@@ -1,55 +1,63 @@
-
-
-### Source configuration
-
-Selection of which tables and columns will be used in ingestion is defined in  `sources`  configuration. </br>
-The structure of configuration part is this:
+This module implements several connectors that can act as a source in either real-time or batch fashion.
+Each of the connectors implements the [`Connector` trait](https://github.com/getdozer/dozer/blob/main/dozer-ingestion/src/connectors/mod.rs) to support being a source to the data pipeline.
+
+## New connector implementation
+
+### Trait
+
+Every connector to external database needs to implement the `Connector` trait [/dozer-ingestion/src/connectors/mod.rs](https://github.com/getdozer/dozer/blob/main/dozer-ingestion/src/connectors/mod.rs)
+
+```rust
+pub trait Connector: Send + Sync + Debug {
+    /// Returns all the external types and their corresponding Dozer types.
+    /// If the external type is not supported, None should be returned.
+    fn types_mapping() -> Vec<(String, Option<FieldType>)>
+    where
+        Self: Sized;
+
+    /// Validates the connector's connection level properties.
+    fn validate_connection(&self) -> Result<(), ConnectorError>;
+
+    /// Lists all the table names in the connector.
+    fn list_tables(&self) -> Result<Vec<TableIdentifier>, ConnectorError>;
+
+    /// Validates the connector's table level properties for each table.
+    fn validate_tables(&self, tables: &[TableIdentifier]) -> Result<(), ConnectorError>;
+
+    /// Lists all the column names for each table.
+    fn list_columns(&self, tables: Vec<TableIdentifier>) -> Result<Vec<TableInfo>, ConnectorError>;
+
+    /// Gets the schema for each table. Only requested columns need to be mapped.
+    ///
+    /// If this function fails at the connector level, such as a network error, it should return a outer level `Err`.
+    /// Otherwise the outer level `Ok` should always contain the same number of elements as `table_infos`.
+    ///
+    /// If it fails at the table or column level, such as a unsupported data type, one of the elements should be `Err`.
+    fn get_schemas(
+        &self,
+        table_infos: &[TableInfo],
+    ) -> Result<Vec<SourceSchemaResult>, ConnectorError>;
+
+    /// Starts outputting data from `tables` to `ingestor`. This method should never return unless there is an unrecoverable error.
+    fn start(&self, ingestor: &Ingestor, tables: Vec<TableInfo>) -> Result<(), ConnectorError>;
+}
+```
+
+Detailed explanation of the data structures and method contracts can be found in [the specification](./SPEC.md).
+
+### Connector functions usage in dozer commands
+
+Dozer uses connector methods in 3 different commands. During `connector ls` execution, dozer just fetches schemas. To get schemas we use `list_tables`, `list_columns` and `get_schemas` methods.
```

Comment:
Are we calling for `get_schemas` during `connector ls` execution?

Reply:
I didn't check the code. The old documentation says so so I included it.

Reply:
oh ok